### ISSUE TITLE
Generic function decorated with functools.cache returns the wrong type

### ISSUE THREAD

**Issue by adonig** [python/cpython#100708]
# Bug report

A generic function decorated with functools.cache returns the wrong type sometimes. Try this example:

```python 
from decimal import Decimal
from functools import cache
from typing import SupportsRound, TypeVar

_T = TypeVar("_T")


@cache
def rnd(val: SupportsRound[_T]) -> _T:
    return round(val, 2)


def test_rnd():
    actual = rnd(Decimal("123"))
    assert actual == Decimal("123")
    assert type(actual) == Decimal
    actual = rnd(123.0)
    assert actual == 123.0
    assert type(actual) == float

```

My understanding is that the test should run without raising any errors but instead the last assertion fails:

```pytb
./xxx/tests/test_cache.py::test_rnd Failed: [undefined]AssertionError: assert <class 'decimal.Decimal'> == float
 +  where <class 'decimal.Decimal'> = type(Decimal('123.00'))
def test_rnd():
        actual = rnd(Decimal("123"))
        assert actual == Decimal("123")
        assert type(actual) == Decimal
        actual = rnd(123.0)
        assert actual == 123.0
>       assert type(actual) == float
E       AssertionError: assert <class 'decimal.Decimal'> == float
E        +  where <class 'decimal.Decimal'> = type(Decimal('123.00'))

xxx/tests/test_cache.py:19: AssertionError
```

# Your environment

- CPython versions tested on: Python 3.11.0
- Operating system and architecture: Darwin mbp 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:54 PDT 2022; root:xnu-8792.41.9~2/RELEASE_X86_64 x86_64

**Comment by adonig** [comment]
I believe this is because functools.cache uses `hash` to [make the cache key](https://github.com/python/cpython/blob/8b1f1251215651c4ef988622345c5cb134e54d69/Lib/functools.py#L441). Maybe it should also hash the argument type.

```
>>> from decimal import Decimal
>>> hash(Decimal("123.0"))
123
>>> hash(123.0)
123
>>> hash(Decimal)
271182336
>>> hash(float)
270673721
```

**Comment by adonig** [comment]
I understand that this is a won't fix and I'm going to close the issue. We ran into this problem because we wrote some code to automatically wrap all the functions in a big module with the cache decorator to speed up a slow calculation. It was a great improvement and some weeks later we suddenly ran into strange errors and had to debug Decimals coming out of nowhere. Now we moved the function out of the module into a utils module so it doesn't get cached and the errors are gone. Sometimes you just wish Python would be a true functional programming language with referential transparency.
